### PR TITLE
Improve Avro GenericRecord and SpecificRecord based row-level extractor performance

### PR DESCRIPTION
**Pull request by jaymo001**
## Description
Improve Avro GenericRecord and SpecificRecord based row-level extractor performance by supporting convert source dataframe into RDD[IndexedRecord] in batch manner.

Major changes:    
1. In SparkRowExtractor.scala, add new extractor method which can be extended to do batch preprocess source dataframe into RDD[IndexRecord].
2. In FeatureTransformation.scala, add logic to extract features from RDD[IndexedRecord].
3. Improve some error messages.

### CODE REVIEW

**Review by xiaoyongzhu — DISMISSED** [review]

File: src/main/scala/com/linkedin/feathr/common/SparkRowExtractor.scala [hunk 1]
```diff
@@ -12,12 +14,27 @@ trait SparkRowExtractor {
...
+   * into specific datatype, e.g. Avro GenericRecord or SpecificRecord.
+   */
+  def hasBatchPreProcessing() = false
+
+  /**
+   * One time batch preprocess the input data source into a RDD[_] for feature extraction later
+   * @param df input data source
+   * @return batch preprocessed dataframe, as RDD[IndexedRecord]
+   */
+  def batchPreProcess(df: DataFrame) : RDD[IndexedRecord] = throw new NotImplementedError("Batch preprocess is not implemented")
```
Comment (bozhonghu) [inline src/main/scala/com/linkedin/feathr/common/SparkRowExtractor.scala:39]: IMO the naming `batchPreProcess` is too general for what this function does. The naming implies that this function can be used for batch preprocessing of data but it is really supposed to be a conversion function between DataFrame -> RDD[IndexdRecord]. Not a blocker but I think something like `convertDfToRdd` would better reflect the purpose of this function.

File: src/main/scala/com/linkedin/feathr/offline/job/FeatureTransformation.scala [hunk 2]
```diff
@@ -673,6 +710,204 @@ private[offline] object FeatureTransformation {
...
+    val transformInfo = featureAnchorWithSources map { featureAnchorWithSource =>
+      val extractor = featureAnchorWithSource.featureAnchor.extractor
+      extractor match {
+        case transformer: AnchorExtractorBase[IndexedRecord] =>
+          // We no longer need prefix for the simplicity of the implementation, instead if there's a feature name
+          // and source data field clash, we will throw exception and ask user to rename the feature.
+          val featureNamePrefix = ""
+          val featureNames = featureAnchorWithSource.selectedFeatures.filter(requestedFeatureNames.contains)
+          val featureNamePrefixPairs = featureNames.map((_, featureNamePrefix))
+          (featureAnchorWithSource, featureNamePrefixPairs, transformer)
```
Comment (bozhonghu) [inline src/main/scala/com/linkedin/feathr/offline/job/FeatureTransformation.scala:798]: Maybe a case class for this?

**Review by bozhonghu — COMMENTED** [review]
Overall LGTM. Just a couple comments on naming and documentation.

File: src/main/scala/com/linkedin/feathr/offline/job/FeatureTransformation.scala [hunk 3]
```diff
@@ -673,6 +710,204 @@ private[offline] object FeatureTransformation {
...
+          s"Invalid extractor $extractor for features:" +
+            s"$requestedFeatureNames requested as tensors")
+    } reduce (_ ++ _)
+    if (logger.isTraceEnabled) {
+      logger.trace(s"Extracted features: $features")
+    }
+
+    /*
+     * Retain feature values for only the requested features, and represent each feature value as a term-vector or as
+     * a tensor, as specified. If tensors are required, create a row for each feature value (that is, the tensor).
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply src/main/scala/com/linkedin/feathr/offline/job/FeatureTransformation.scala]: Updated.

File: src/main/scala/com/linkedin/feathr/common/SparkRowExtractor.scala (hunk 1 above)
Comment (jaymo001) [reply src/main/scala/com/linkedin/feathr/common/SparkRowExtractor.scala]: Changed to convertToAvroRdd

**Comment by xiaoyongzhu** [comment]
Seems the sbt test is failed. Not sure if it's related with this change.

File: src/main/scala/com/linkedin/feathr/offline/anchored/keyExtractor/MVELSourceKeyExtractor.scala [hunk 4]
```diff
@@ -86,5 +86,6 @@ private[feathr] class MVELSourceKeyExtractor(val anchorExtractorV1: AnchorExtrac
   // this helps to reduce the number of joins
   // to the observation data
   // The default toString does not work, because toString of each object have different values
-  override def toString: String = getClass.getSimpleName + " with keyExprs:" + keyExprs.mkString(" key:")
+  override def toString: String = getClass.getSimpleName + " with keyExprs:" + keyExprs.mkString(" key:") +
+    anchorExtractorV1.toString
```
Comment (d4ve) [inline src/main/scala/com/linkedin/feathr/offline/anchored/keyExtractor/MVELSourceKeyExtractor.scala:90]: This change is a bit hard to understand. Why would we just append the anchorExtractorV1.toString to the end like this?

File: src/main/scala/com/linkedin/feathr/offline/anchored/keyExtractor/MVELSourceKeyExtractor.scala [hunk 5]
```diff
@@ -86,5 +86,6 @@ private[feathr] class MVELSourceKeyExtractor(val anchorExtractorV1: AnchorExtrac
   // this helps to reduce the number of joins
   // to the observation data
   // The default toString does not work, because toString of each object have different values
```
Comment (d4ve) [inline src/main/scala/com/linkedin/feathr/offline/anchored/keyExtractor/MVELSourceKeyExtractor.scala:88]: Side comment: This is beyond this scope of this change, but it looks dubious to be using .toString to drive behavior like this. There should have been another function added, to drive the behavior described here. (Is this the reason you needed to add something to the .toString implementation below ...?)

File: src/main/scala/com/linkedin/feathr/offline/job/FeatureTransformation.scala [hunk 6]
```diff
@@ -1,7 +1,9 @@
 package com.linkedin.feathr.offline.job
 
-import com.linkedin.feathr.common._
 import com.linkedin.feathr.common.exception.{ErrorLabel, FeathrException, FeathrFeatureTransformationException}
+import com.linkedin.feathr.common.tensor.TensorData
+import com.linkedin.feathr.common.types.FeatureType
+import com.linkedin.feathr.common.{AnchorExtractorBase, _}
```
Comment (d4ve) [inline src/main/scala/com/linkedin/feathr/offline/job/FeatureTransformation.scala:6]: Are the changes in this file newly added? Or just being transferred over from the LinkedIn fork?

File: src/main/scala/com/linkedin/feathr/common/AnchorExtractor.scala [hunk 7]
```diff
@@ -39,12 +37,14 @@ trait AnchorExtractor[T] extends AnchorExtractorBase[T] with SparkRowExtractor {
...
 
   /**
    * Get the feature value from the row
    * @param datum input row
    * @return A map of feature name to feature value
    */
-  def getFeaturesFromRow(datum: GenericRowWithSchema): Map[String, FeatureValue] = getFeatures(datum.asInstanceOf[T])
+  def getFeaturesFromRow(datum: Any): Map[String, FeatureValue] = getFeatures(datum.asInstanceOf[T])
+
+  override def toString: String = getClass.getSimpleName
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply src/main/scala/com/linkedin/feathr/common/AnchorExtractor.scala:49]: Feathr relies on toString to identify group of extractors. We should create an identifier method for the purpose instead. This could be done cleanly after the FCM release.

File: src/main/scala/com/linkedin/feathr/common/SparkRowExtractor.scala [hunk 8]
```diff
@@ -12,12 +14,27 @@ trait SparkRowExtractor {
...
+   * into specific datatype, e.g. Avro GenericRecord or SpecificRecord.
+   */
+  def isLowLevelRddExtractor() = false
+
+  /**
+   * One time batch preprocess the input data source into a RDD[IndexedRecord] for feature extraction later
+   * @param df input data source
+   * @return batch preprocessed dataframe, as RDD[IndexedRecord]
+   */
+  def convertToAvroRdd(df: DataFrame) : RDD[IndexedRecord] = throw new NotImplementedError("Batch preprocess is not implemented")
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply src/main/scala/com/linkedin/feathr/common/SparkRowExtractor.scala]: Create trait WorkWithAvroRdd instead.

File: src/main/scala/com/linkedin/feathr/offline/job/FeatureTransformation.scala (hunk 6 above)
Comment (jaymo001) [reply src/main/scala/com/linkedin/feathr/offline/job/FeatureTransformation.scala:6]: There's minor refactor based off Li fork.

**Comment by xiaoyongzhu** [comment]
Thanks, this PR looks good to me. I also like the added unit test to make sure the parsing works correctly.

Have a few minor comments, and also let's make sure it works when reading features from the registry (that's where most of the edge cases are from).

File: src/main/scala/com/linkedin/feathr/common/AnchorExtractor.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (d4ve) [reply src/main/scala/com/linkedin/feathr/common/AnchorExtractor.scala:49]: It's important that this requirement be documented in the scaladoc here, or in the parent class, depending on how general this requirement is.

File: src/main/scala/com/linkedin/feathr/offline/anchored/keyExtractor/MVELSourceKeyExtractor.scala (hunk 4 above)
Comment (d4ve) [reply src/main/scala/com/linkedin/feathr/offline/anchored/keyExtractor/MVELSourceKeyExtractor.scala:90]: At a minimum, should the toString be modified to make the output more legible? E.g. adding `"anchorExtractor: "` to the string before `anchorExtractorV1.toString`?

File: src/main/scala/com/linkedin/feathr/common/WorkWithAvroRdd.scala [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+ * convertToAvroRdd will be called by Feathr engine before calling getKeyFromRow() and getFeaturesFromRow() in AnchorExtractor.
+ */
+trait WorkWithAvroRdd {
+
+  /**
+   * One time batch converting the input data source into a RDD[IndexedRecord] for feature extraction later
+   * @param df input data source
+   * @return batch preprocessed dataframe, as RDD[IndexedRecord]
+   */
+  def convertToAvroRdd(df: DataFrame) : RDD[IndexedRecord] = throw new UnsupportedOperationException("Batch preprocess is not implemented")
```
[the remark this answers is not in the document]
Comment (jaymo001) [reply src/main/scala/com/linkedin/feathr/common/WorkWithAvroRdd.scala]: Renamed to CanConvertToAvroRDD

**Review by bozhonghu — APPROVED** [review]
LGTM!

**Review by d4ve — APPROVED** [review]
